test/test_logging.cpp

```diff
@@ -485,3 +487,52 @@ TEST(TestLogging, test_root_logger_after_nonexistent)
     rcutils_logging_set_logger_level(
       "", original_root_severity));
 }
+
+TEST(TestLogging, test_logging_macro_thread_safety)
+{
+  // This tests whether or not using logging macros from multiple threads is safe or not.
+
+  // This test is based on an issue found in the optimization of the logging macros,
+  // and therefore has a very specific trigger scenario, which is described more
+  // in the steps below.
+  // See: https://github.com/ros2/rcutils/pull/393
+
+  // This test is likely to be flakey false-positive, meaning it's possible that
+  // it will pass even if the macros are not thread-safe and may require running
+  // repeatedly to detect problems.
+
+  ASSERT_EQ(RCUTILS_RET_OK, rcutils_logging_initialize());
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    EXPECT_EQ(RCUTILS_RET_OK, rcutils_logging_shutdown());
+  });
+
+  // One of the loggers needs to be set "by the user" to trigger the optimization.
+  ASSERT_EQ(RCUTILS_RET_OK, rcutils_logging_set_logger_level("", RCUTILS_LOG_SEVERITY_DEBUG));
+
+  // In threads, in a loop do a log call on many different logger names.
+  // The message doesn't matter.
+  std::size_t loop_count = 10;
+  auto task = [&loop_count](std::size_t thread_number) {
+      for (std::size_t i = 0; i < loop_count; ++i) {
+        rcutils_log_location_t location = {"func", "file", 42u};
+        rcutils_log(
```

Comment:
`rcutils_log` is not thread-safe, this is not allowed to call with multi-thread? saying we are testing something `not supported`?

https://github.com/ros2/rcutils/blob/3485f1bdc805ead1e9c26509479dc8c1f94b6951/include/rcutils/logging.h#L480-L505

Reply:
Yes, it is tricky, and the reason that I introduced the problem to begin with.

It is definitely the case that `rcutils_log` is *not* thread-safe with respect to `rcutils_set_logger_level` (before or after this PR).  If you call the two of those from two different threads (without any additional locking), the results will be undefined.

However, prior to my optimizations, `rcutils_log` was thread-safe with itself.  That is, because `rcutils_log` was only reading data out of the underlying data structures, it didn't matter how many threads were calling it.  With the current code (before this PR), that is no longer the case because the individual threads can write to the hashmap, leading to corruption/crashing.  This PR fixes it by essentially making `rcutils_log` completely read-only again.

One other bit of information here pulls this all together.  If we are coming via the `RCLCPP_` logging macros, there is actually no problem; `rclcpp` has a global lock around all logging.  However, some of our lower-level components (like `rcl`, `rmw`, etc) use the `RCUTILS_` macros, and they do not have this global lock.

I think we should consider adding some kind of reader-writer lock here, which would solve most of the problem.  However, we have classically not done any locking in the rcutils layer, so we would have to change that policy (and figure out how to do it in a cross-platform way).

Does all of that make sense @fujitatomoya ?

Reply:
The problem is that this is just not how people use the logging macros, which of course use this function.

Also, there's no good place to inject this locking other than in rcutils, and we've been trying to avoid using threading in rcutils/rmw/rcl in order to use client library language specific mechanisms for threading when possible.

I think a reader-writer lock is perfect for this case and I think we could probably implement a relatively efficient (compared to actually locking with a mutex) lock-free form of this, since the writer lock is only needed when setting log levels and that should be done infrequently. It's something I intend to investigate.

Also, I think this doc block was added after the functions were written and just didn't consider how it was being used. Unfortunately, I think there will be many such inaccuracies in the doc blocks that were back filled a few years ago in the REP-2004 QD effort.

Reply:
understood, thanks for the explanation guys! i am good to go.